# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: layout/style/FontFaceSetDocumentImpl.h
# Commit: 22f567c00af3
# Full Hash: 22f567c00af321b797590746f181763abb07d83b
# Author: Andrew Osmond <aosmond@mozilla.com>
# Date: 2022-06-29 03:44:30
# Regressor Bug: 1771493
# File Overlap Count: 1
# Description:
#   Bug 1771493 - Part 4. Split FontFaceSetDocumentImpl from FontFaceSetImpl. r=emilio
#   
#   This patch splits document specific functionality from FontFaceSetImpl
#   into a new class FontFaceSetDocumentImpl. This will make it easier to
#   fork FontFaceSetImpl for workers.
# ==============================================================================

diff -r 609a5d93bf08 -r 22f567c00af3 layout/style/FontFaceSetDocumentImpl.h
--- /dev/null	Thu Jan 01 00:00:00 1970 +0000
+++ b/layout/style/FontFaceSetDocumentImpl.h	Tue Jun 28 15:58:11 2022 +0000
@@ -0,0 +1,115 @@
+/* -*- Mode: C++; tab-width: 8; indent-tabs-mode: nil; c-basic-offset: 2 -*- */
+/* vim: set ts=8 sts=2 et sw=2 tw=80: */
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#ifndef mozilla_dom_FontFaceSetDocumentImpl_h
+#define mozilla_dom_FontFaceSetDocumentImpl_h
+
+#include "mozilla/dom/FontFaceSetImpl.h"
+#include "nsICSSLoaderObserver.h"
+#include "nsIDOMEventListener.h"
+
+namespace mozilla::dom {
+
+class FontFaceSetDocumentImpl final : public FontFaceSetImpl,
+                                      public nsIDOMEventListener,
+                                      public nsICSSLoaderObserver {
+  NS_DECL_ISUPPORTS_INHERITED
+
+ public:
+  NS_DECL_NSIDOMEVENTLISTENER
+
+  FontFaceSetDocumentImpl(FontFaceSet* aOwner, dom::Document* aDocument);
+
+  void Initialize();
+  void Destroy() override;
+
+  dom::Document* Document() const override { return mDocument; }
+
+  // gfxUserFontSet
+
+  nsresult StartLoad(gfxUserFontEntry* aUserFontEntry,
+                     uint32_t aSrcIndex) override;
+
+  bool IsFontLoadAllowed(const gfxFontFaceSrc&) override;
+
+  nsPresContext* GetPresContext() const override;
+
+  bool UpdateRules(const nsTArray<nsFontFaceRuleContainer>& aRules) override;
+
+  RawServoFontFaceRule* FindRuleForEntry(gfxFontEntry* aFontEntry) override;
+
+  /**
+   * Notification method called by the nsPresContext to indicate that the
+   * refresh driver ticked and flushed style and layout.
+   * were just flushed.
+   */
+  void DidRefresh() override;
+
+  // nsICSSLoaderObserver
+  NS_IMETHOD StyleSheetLoaded(StyleSheet* aSheet, bool aWasDeferred,
+                              nsresult aStatus) override;
+
+  // For ServoStyleSet to know ahead of time whether a font is loadable.
+  void CacheFontLoadability() override;
+
+  void EnsureReady() override;
+
+  bool Add(FontFaceImpl* aFontFace, ErrorResult& aRv) override;
+
+  void FlushUserFontSet() override;
+  void MarkUserFontSetDirty() override;
+
+ private:
+  ~FontFaceSetDocumentImpl() override;
+
+  uint64_t GetInnerWindowID() override;
+
+  void RemoveDOMContentLoadedListener();
+
+  nsresult CreateChannelForSyncLoadFontData(
+      nsIChannel** aOutChannel, gfxUserFontEntry* aFontToLoad,
+      const gfxFontFaceSrc* aFontFaceSrc) override;
+
+  // search for @font-face rule that matches a userfont font entry
+  RawServoFontFaceRule* FindRuleForUserFontEntry(
+      gfxUserFontEntry* aUserFontEntry) override;
+
+  void FindMatchingFontFaces(const nsTHashSet<FontFace*>& aMatchingFaces,
+                             nsTArray<FontFace*>& aFontFaces) override;
+
+  void InsertRuleFontFace(FontFaceImpl* aFontFace, FontFace* aFontFaceOwner,
+                          StyleOrigin aOrigin,
+                          nsTArray<FontFaceRecord>& aOldRecords,
+                          bool& aFontSetModified);
+
+  // Helper function for HasLoadingFontFaces.
+  void UpdateHasLoadingFontFaces() override;
+
+  /**
+   * Returns whether there might be any pending font loads, which should cause
+   * the mReady Promise not to be resolved yet.
+   */
+  bool MightHavePendingFontLoads() override;
+
+#ifdef DEBUG
+  bool HasRuleFontFace(FontFaceImpl* aFontFace);
+#endif
+
+  already_AddRefed<gfxFontSrcPrincipal> CreateStandardFontLoadPrincipal()
+      const override;
+
+  TimeStamp GetNavigationStartTimeStamp() override;
+
+  // The document this is a FontFaceSet for.
+  RefPtr<dom::Document> mDocument;
+
+  // The @font-face rule backed FontFace objects in the FontFaceSet.
+  nsTArray<FontFaceRecord> mRuleFaces;
+};
+
+}  // namespace mozilla::dom
+
+#endif  // !defined(mozilla_dom_FontFaceSetDocumentImpl_h)